### ISSUE TITLE
Remove redundant TRANSPOSE Ops generated by onnx-tf

### ISSUE THREAD

**Issue by jinevening** [Samsung/ONE#5401]
## What

Remove redundant TRANSPOSE Ops generated by onnx-tf

## Why

We're going to support ONNX model in ONE via the following steps.

ONNX -> TF (.pb) -> TFLite (.tflite) -> Circle (.circle)

[onnx-tensorflow](https://github.com/onnx/onnx-tensorflow) will be used to convert ONNX file to TF (.pb) file. It generates many Transpose Ops to convert the data layout from NCHW (used by ONNX) to NHWC (used by TF). These Transpose Ops are indeed meaningless in terms of DNN behavior, but introduce a huge overhead.

## How

To remove these Transpose Ops, we'll change NCHW Ops to NHWC Ops and use [RemoveRedundantTransposePass](https://github.com/Samsung/ONE/pull/5172). For example,

Before: NCHW tensor -> **Op (NCHW)** -> NCHW tensor

After: NCHW tensor -> **Transpose (perm:[0,2,3,1]) -> NHWC tensor -> Op (NHWC) -> NHWC tensor -> Transpose (perm:[0,3,1,2])** -> NCHW tensor

`onnx-tensorflow` does this transformation only for special Ops (e.g., Conv, TConv, etc). We're going to do this transformation for all Ops to make `RemoveRedundantTransposePass` remove redundant Transpose Ops.

For example, let's assume the above transformation is done for two subsequent Ops

NCHW tensor -> Transpose (perm:[0,2,3,1]) -> NHWC tensor -> Op (changed to use NHWC) -> NHWC tensor -> **Transpose (perm:[0,3,1,2]) -> NCHW tensor -> Transpose (perm:[0,2,3,1])** -> NHWC tensor -> Op (changed to use NHWC) -> NHWC tensor -> Transpose (perm:[0,3,1,2]) -> NCHW tensor

The bold Transpose Ops can be removed by `RemoveRedundantTransposePass` as they are redundant.

For this, we need a pass (`ConvertNCHWToNHWCPass`) that updates NCHW Ops to NHWC Ops.

## Status

- [x] Convert NCHW to NHWC #5438 @jinevening 
- [x] Remove redundant transpose #5437 @struss 
- [x] Remove redundant split #5456 @struss 
- [x] Remove redundant slice #5416 @struss 
- [x] Convert Transpose to Reshape #5703 @struss 
- [x] Remove redundant Reshape #5275 @struss

**Comment by jinevening** [comment]
I'm working on draft. MUL is transformed as below.

![image](https://user-images.githubusercontent.com/5449554/102564406-b9fb2380-411e-11eb-8ff5-92ad3a7c3e66.png)

ADD and PAD are transformed similarly.

Transpose Ops between the converted Ops can be removed by `RemoveRedundantTransposePass`. The graph applied with `RemoveRedundantTransposePass` is shown below.

![image](https://user-images.githubusercontent.com/5449554/102565727-65a57300-4121-11eb-86a6-6c407d88fe4a.png)

TODO. Improve `RemoveRedundantTransposePass` to cover the below case. @struss Can you help with this?

Before
```
                     Op1
                      |
                  Transpose (0,3,1,2)
                   /     \
Transpose (0,2,3,1)    Transpose (0,2,3,1)
  |                       |
 Op2                      Op3
```
After
```
                     Op1
                   /     \
                Op2       Op3
```

**Comment by struss** [comment]
> TODO. Improve `RemoveRedundantTransposePass` to cover the below case. @struss Can you help with this?
> 
> Before
> 
> ```
>                      Op1
>                       |
>                   Transpose (0,3,1,2)
>                    /     \
> Transpose (0,2,3,1)    Transpose (0,2,3,1)
>   |                       |
>  Op2                      Op3
> ```
> 
> After
> 
> ```
>                      Op1
>                    /     \
>                 Op2       Op3
> ```

@jinevening Okay I will update `RemoveRedundantTransposePass` to cover this case.

**Comment by struss** [comment]
Current logic only handle this case below
```
  Op1
   |
Transpose 
   | (with only one passing)
 Transpose  
   | 
(any number of succs)
```
to change to reuse first node below.
```
  Op1
   |
  Transpose 
   |       \
   |        \
 Transpose   |
             |
(any number of succs)
```
I think change logic to create new transpose.

**Comment by jinevening** [comment]
Another issue: Input and Output

`onnx-tf` inserts Transpose Ops at the beginning and at the end of the model to convert NCHW to NHWC.

![image](https://user-images.githubusercontent.com/5449554/102730194-240e0580-4377-11eb-9d68-60bf9c3a9b45.png)

`ConvertNCHWToNHWCPass` will change the input shape to NHWC and remove the Transpose Ops.

**Comment by jinevening** [comment]
Another issue.

onnx-tf inserts RESHAPE instead of TRANSPOSE if possible.

![image](https://user-images.githubusercontent.com/5449554/102760549-02ce0900-43b9-11eb-9b31-1fbaf43226ab.png)

Reshape->Transpose in the graph is an identity function, so we can remove them. But this pattern is not captured by RemoveRedundantTransposePass.

**Comment by jinevening** [comment]
@struss Can you extend `RemoveRedundantTransposePass` to cover the [above case](https://github.com/Samsung/ONE/issues/5401#issuecomment-748886742) as well (after #5437 is merged) ?

I think it would be better to handle the case in `RemoveRedundantTransposePass` rather than making another pass.

**Comment by struss** [comment]
> @struss Can you extend `RemoveRedundantTransposePass` to cover the [above case](https://github.com/Samsung/ONE/issues/5401#issuecomment-748886742) as well (after #5437 is merged) ?

Okay I will do this.

> I think it would be better to handle the case in `RemoveRedundantTransposePass` rather than making another pass.

then I think I need to rename `RemoveRedundantTransposePass` to something new.

**Comment by struss** [comment]
@jinevening I think we need to consider reverse order case also to be handled.
```
CircleNode -> CircleTranspose -> CircleReshape -> CircleNode
```
I think it will also need to be handled.
so, what about adding a pass for `Transpose to Reshape` or `Reshape to Transpose`?

**Comment by jinevening** [comment]
Reshape is cheaper than Transpose in general, so it would be better to transform Transpose to Reshape and remove redundant Reshape Ops (You've already made a draft! #5266).

**Comment by struss** [comment]
I have a little question. if `Transpose to Reshape` added, then I think this pass must run after [this](https://github.com/Samsung/ONE/pull/5517) as it could change `Transpose` that removed.
example below,
```
Input (1x128x32x1) -> Transpose (1x1x128x32) -> Add (1x1x128x32) -> Transpose (1x128x32x1) -> Output (1x128x32x1)
* Each values in () means output shape of operation.
```
In this example, the best result is
```
Input (1x128x32x1) -> Add (1x128x32x1) -> Output (1x128x32x1)
* with Add const also converted.
```
but If `Transpose to Reshape` activate first:
```
Input (1x128x32x1) -> Reshape (1x1x128x32) -> Add (1x1x128x32) -> Reshape (1x128x32x1) -> Output (1x128x32x1)
```
Is there any way to make on pass activate before other?

**Comment by seanshpark** [comment]
All tasks seems to be finished. Can we close this issue? or maybe add more tasks?

**Comment by jinevening** [comment]
> All tasks seems to be finished. Can we close this issue? or maybe add more tasks?

Let's close this issue.
